repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/KeystoneRequestHandler.scala

```diff
@@ -0,0 +1,287 @@
+
+import java.io.InputStream
+import java.util.{Calendar, GregorianCalendar}
+import javax.servlet.http.HttpServletResponse._
+import javax.ws.rs.core.MediaType
+
+import com.fasterxml.jackson.core.JsonProcessingException
+import com.typesafe.scalalogging.slf4j.LazyLogging
+import org.apache.http.HttpHeaders
+import org.apache.http.client.utils.DateUtils
+import org.joda.time.format.ISODateTimeFormat
+import org.openrepose.commons.utils.http.{CommonHttpHeader, ServiceClientResponse}
+import org.openrepose.core.services.serviceclient.akka.AkkaServiceClient
+import play.api.libs.functional.syntax._
+import play.api.libs.json.Reads._
+import play.api.libs.json._
+
+import scala.collection.JavaConverters._
+import scala.io.Source
+import scala.util.{Failure, Success, Try}
+
+/**
+ * Contains the functions which interact with the Keystone API.
+ */
+class KeystoneRequestHandler(identityServiceUri: String, akkaServiceClient: AkkaServiceClient, traceId: Option[String])
+  extends LazyLogging {
+
+  import KeystoneRequestHandler._
+
+  /**
+   * Call to the Keystone service to get the admin token
+   * @return a Successful token, or a Failure
+   */
+  final def getAdminToken(adminUsername: String, adminPassword: String): Try[String] = {
+    //authenticate, or get the admin token
+    val authenticationPayload = Json.obj(
+      "auth" -> Json.obj(
+        "passwordCredentials" -> Json.obj(
+          "username" -> adminUsername,
+          "password" -> adminPassword
+        )
+      )
+    )
+
+    val akkaResponse = Try(akkaServiceClient.post(ADMIN_TOKEN_KEY,
+      s"$identityServiceUri$TOKEN_ENDPOINT",
+      (Map(CommonHttpHeader.ACCEPT.toString -> MediaType.APPLICATION_JSON)
+        ++ traceId.map(CommonHttpHeader.TRACE_GUID.toString -> _)).asJava,
+      Json.stringify(authenticationPayload),
+      MediaType.APPLICATION_JSON_TYPE
+    ))
+
+    akkaResponse match {
+      case Success(serviceClientResponse) =>
+        serviceClientResponse.getStatus match {
+          case statusCode if statusCode >= 200 && statusCode < 300 =>
+            val jsonResponse = Source.fromInputStream(serviceClientResponse.getData).getLines().mkString("")
+            val json = Json.parse(jsonResponse)
+            Try(Success((json \ "access" \ "token" \ "id").as[String])) match {
+              case Success(s) => s
+              case Failure(f) =>
+                Failure(IdentityCommunicationException("Token not found in identity response during admin authentication", f))
+            }
+          case SC_REQUEST_ENTITY_TOO_LARGE | SC_TOO_MANY_REQUESTS =>
+            Failure(OverLimitException(buildRetryValue(serviceClientResponse), "Rate limited when getting admin token"))
+          case statusCode if statusCode >= 500 =>
+            Failure(IdentityCommunicationException("Identity Service not available to get admin token"))
+          case _ => Failure(new Exception("Unable to successfully get admin token from Identity"))
```

Comment:
what's wrong with using IdentityCommunicationException? is it not an accurate description of what happened?